core/src/main/java/com/linecorp/armeria/internal/client/PublicSuffix.java

```diff
@@ -0,0 +1,172 @@
+                String line;
+                while ((line = reader.readLine()) != null) {
+                    final String[] labels = line.split("\\.");
+                    TrieNode node = trie;
+                    for (int i = labels.length - 1; i >= 0; i--) {
+                        // assume wildcard is at the first position, we don't need to create a new node for it
+                        // but set the current node's isWildcard = true instead
+                        if (labels[i].equals("*")) {
+                            node.isWildcard = true;
+                            break;
+                        }
+                        if (node.children == null) {
+                            node.children = new HashMap<>();
+                        }
+                        if (node.children.containsKey(labels[i])) {
+                            node = node.children.get(labels[i]);
+                        } else {
+                            final TrieNode newNode = new TrieNode();
+                            if (labels[i].charAt(0) == '!') {
+                                newNode.isException = true;
+                                node.children.put(labels[i].substring(1), newNode);
+                            } else {
+                                node.children.put(labels[i], newNode);
+                            }
+                            node = newNode;
+                        }
+                    }
+                    node.isEnd = true;
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    /**
+     * Returns whether a domain is a public suffix. Suffix matching is done as follow:
+     * <ol>
+     *     <li>Split the domain by dot '.' into labels. Leading and trailing dots are ignored.</li>
+     *
+     *     <li>Iterate from the last label to first, at the same time traverse the trie from root. At index i,
+     *     current node's children should contain a node with key = labels[i]. If not, then the domain does
+     *     not match any rule and thus is not a public suffix.</li>
+     *
+     *     <li>If the node has {@code isWildcard = true}, it can match any label at index i - 1. Since the
+     *     wildcard is always the leftmost label, i should be 1, otherwise the domain does not match. For
+     *     example, {@code *.ck} matches {@code a.ck}, but not {@code a.b.ck}.</li>
+     *
+     *     <li>If we have a wildcard match, we still need to check for exception. If label i - 1 matches the
+     *     next node and that node has {@code isException = true}, the wildcard is rejected.</li>
+     * </ol>
+     *
+     * @param domain A canonicalized domain. An internationalized domain name (IDN) should be Punycode
+     *               encoded, for example, using {@link java.net.IDN#toASCII(String)}.
+     */
+    public boolean isPublicSuffix(String domain) {
+        final String[] labels = domain.split("\\.");
```

Comment:
Could you nomarlize the dot-like characters like Guava did?
https://github.com/google/guava/blob/a6752d73acba20295ecdcddd959dac904dbfa4d3/guava/src/com/google/common/net/InternetDomainName.java#L135-L138

Reply:
Normalization could be helpful in domain matching too, we're also checking dots there. Should we do it earlier, maybe at `ensureDomainAndPath`?

Reply:
Actually, there are things I think we could lift up the chain, normalization and procesing leading/trailing dots in cookie domain. Leading dot should be ignored, and trailing dot, actually as I reread it, should cause the client to ignore the attribute completely https://tools.ietf.org/html/rfc6265#section-4.1.2.3. This better be set at cookie creation, I think? An addition to `CookieBuilder` would be nice.

Reply:
> Should we do it earlier, maybe at ensureDomainAndPath?

I think we need to normalize here because this method is called by `CookiePolicy.domainMatches(...)`, which is a public API.

> Leading dot should be ignored, and trailing dot, actually as I reread it, should cause the client to ignore the attribute completely tools.ietf.org/html/rfc6265#section-4.1.2.3. This better be done at cookie creation, I think? An addition to CookieBuilder would be nice.

I think it would be nice to add normalizing domain to `CookieBuilder`. Additionally, we can raise `IllegalArgumentException` if a `domain` attribute is a public suffix.

Reply:
Looks like cookie domain is already validated for unusual characters in Builder. Probably don't need to normalize here anymore I think? https://github.com/line/armeria/blob/2a27bd264dceb072a2323bc5aac497d5d2218dd4/core/src/main/java/com/linecorp/armeria/common/CookieBuilder.java#L42

Reply:
I guess what @ikhoon says is that anyone can call this API with any invalid value via `CookiePolicy.domainMatches()`.

Maybe we could optimize the validation as suggested by @tumile, i.e. validate and normalize more aggressively in `CookieBuilder.domain(String)`, if:

- we hide `domainMatches()` somehow from the public API; or
- we change `domainMatches()` accept a `Cookie` instead of a `String`.

Makes sense?

Reply:
I added the dots check into `CookieBuilder`. `domainMatches` is fairly simple so I made it private in both `AcceptOriginPolicy` and `CookieJar`. There's minor difference, like domain matching in `CookieJar` doesn't have to check for null domains, so I think that justifies the duplication.